crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -1230,6 +1234,135 @@ impl ParallelSparseTrie {
+                        SparseSubtrieType::from_path(&node_b.path)
+                })
+                .collect();
+
+            // Take the lower subtries in the same order that the nodes were grouped into, so that
+            // the two can be zipped together. This also must be collected into a Vec for rayon's
+            // `zip` to be happy.
+            let lower_subtries: Vec<_> = node_groups
+                .iter()
+                .map(|nodes| {
+                    let node = nodes.first().expect("no empty groups");
+                    let Some(idx) = SparseSubtrieType::from_path(&node.path).lower_index() else {
+                        panic!("upper subtrie node {node:?} found amongst lower nodes");
+                    };
+                    // due to the nodes being sorted secondarily on their path, and chunk_by keeping
+                    // the first element of each group, the `path` here will necessarily be the
+                    // shortest path being revealed for each subtrie. Therefore we can reveal the
+                    // subtrie itself using this path and retain correct behavior.
+                    self.lower_subtries[idx].reveal(&node.path);
+                    (idx, self.lower_subtries[idx].take_revealed().expect("just revealed"))
+                })
+                .collect();
+
+            let (tx, rx) = mpsc::channel();
+
+            // Zip the lower subtries and their corresponding node groups, and reveal lower subtrie
+            // nodes in parallel
+            lower_subtries
+                .into_par_iter()
+                .zip(node_groups.into_par_iter())
+                .map(|((subtrie_idx, mut subtrie), nodes)| {
+                    // reserve space in the HashMap ahead of time; doing it on a node-by-node basis
+                    // can cause multiple re-allocations as the hashmap grows.
+                    subtrie.nodes.reserve(nodes.len());
+
+                    for node in nodes {
+                        // Reveal each node in the subtrie, returning early on any errors
+                        let res = subtrie.reveal_node(node.path, &node.node, node.masks.clone());
+                        if res.is_err() {
+                            return (subtrie_idx, subtrie, res)
+                        }
+                    }
+                    (subtrie_idx, subtrie, Ok(()))
+                })
+                .for_each_init(|| tx.clone(), |tx, result| tx.send(result).unwrap());
+
+            drop(tx);
+
+            let mut any_err = Ok(());
+            for (subtrie_idx, subtrie, res) in rx {
+                self.lower_subtries[subtrie_idx] = LowerSparseSubtrie::Revealed(subtrie);
+                if res.is_err() {
+                    any_err = res;
+                }
+            }
+
+            // TODO we're just returning the last seen error here, not sure if there's a more
+            // idiomatic thing to do...
+            any_err
+        }
```

Comment:
i would probably just loop through the `rx` results here and if we get an err, return early (maybe just by doing `res?`). and the fn would have Ok(()) at the end

Reply:
I see, that seems fine then